astropy/io/fits/hdu/compressed/compressed.py

```diff
@@ -539,12 +563,12 @@ def _update_header_data(
             self.name = name
 
     def _scale_data(self, data):
-        if self._orig_bzero != 0 or self._orig_bscale != 1:
+        if self._orig_bzero != 0 or self._orig_bscale != 1 or self._blank is not None:
             new_dtype = self._dtype_for_bitpix()
             data = np.array(data, dtype=new_dtype)
 
-            if "BLANK" in self._header:
-                blanks = data == np.array(self._header["BLANK"], dtype="int32")
+            if self._blank is not None:
+                blanks = data == np.array(self._blank, dtype="int32")
```

Comment:
I agree, this is very weird. I haven't changed this part of the code (except to execute it when `BLANK` is set, even without `BSCALE/BZERO`).  I think it would make more sense to do the comparison of data to the blank value before the conversion to float. I don't have any insight into the reason for using `int32` for the datatype.

Maybe this is related to the `uint` thing, which is a can of worms.  I see that `_dtype_for_bitpix()` returns a `uint` type for the special values of `bscale=1` and `bzero=2**bits-1`.  So the type conversion may be changing `data` from `int` to `uint` (rather than to `float`).  The `data` array will stay as type `uint` unless there are blanks, in which case the `np.where(blanks, np.nan, data)` will convert it to a float type.  (I did test this to make sure it works.)

It appears to me this code is simply wrong for `uint` datatypes.  The comparison to the `BLANK` value needs to be done **before the conversion to `uint`**.  If it is done after the conversion to `uint` the results will _not_ be correct.

I have not looked at the `uint` part of the code at all.  I will probably try to fix this since I need a version of this code that works now (for access to Pan-STARRS images, which die with an exception). But if it does not get integrated into a release before #15474 is committed, it will doubtless have to be reworked after that.  I imagine the handling of `BLANK` keywords for unsigned integers in `CompImageHDU` is correct (at least I hope so!) and is different than the code here.

Reply:
> It appears to me this code is simply wrong for `uint` datatypes. The comparison to the `BLANK` value needs to be done **before the conversion to `uint`**. If it is done after the conversion to `uint` the results will _not_ be correct.

You're right, I did not even notice although I quoted the comparison including `BZERO` elsewhere – but clearly the sensible way to do this would be comparing before conversion. Plain bug to me.